typescript/core/src/ChunkCursor.ts

```diff
@@ -211,80 +150,78 @@ export class ChunkCursor {
+      return;
     }
 
-    const cursor = this.#messageIndexCursors.peek();
-    if (!cursor) {
+    this.#orderedMessageOffset.sort(([logTimeA], [logTimeB]) => Number(logTimeA - logTimeB));
+    if (reverse) {
+      // If we used `logTimeB - logTimeA` as the comparator for reverse iteration, messages with
+      // the same timestamp would not be in reverse order. To avoid this problem we use reverse()
+      // instead.
+      this.#orderedMessageOffset.reverse();
+    }
+
+    const [logTimeFirstMessage] = this.#orderedMessageOffset[0]!;
+    if (logTimeFirstMessage < this.chunkIndex.messageStartTime) {
       throw new Error(
-        `Unexpected empty cursor for chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+        `Chunk contains a message with logTime (${logTimeFirstMessage}) earlier than chunk messageStartTime (${this.chunkIndex.messageStartTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
       );
     }
 
-    return cursor.records[cursor.index]![0];
+    const [logTimeLastMessage] = this.#orderedMessageOffset[this.#orderedMessageOffset.length - 1]!;
+    if (logTimeLastMessage > this.chunkIndex.messageEndTime) {
+      throw new Error(
+        `Chunk contains a message with logTime with logTime (${logTimeLastMessage}) later than chunk messageEndTime (${this.chunkIndex.messageEndTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+      );
+    }
+
+    // Remove message offsets whose log time is not within [startTime, endTime];
+    let startIndex: number | undefined;
+    let endIndex: number | undefined;
+    const startTime = reverse ? this.#endTime : this.#startTime;
+    const endTime = reverse ? this.#startTime : this.#endTime;
+
+    const iteratee = (logTime: bigint) => (reverse ? -logTime : logTime);
+    if (startTime != undefined) {
+      startIndex = sortedIndexBy(this.#orderedMessageOffset, startTime, iteratee);
+    }
+    if (endTime != undefined) {
+      endIndex = sortedIndexBy(this.#orderedMessageOffset, endTime, iteratee);
+      // sortedIndexBy returns the minimum index but for the end index we actually want the highest index since
+      // endTime is inclusive. So we count up the endIndex manually until we reach a logTime that is not included anymore.
+      while (endIndex < this.#orderedMessageOffset.length) {
+        const logTime = this.#orderedMessageOffset[endIndex]![0];
+        if (reverse ? logTime < endTime : logTime > endTime) {
+          break;
+        }
+        endIndex++;
+      }
+    }
+
+    if (startIndex != undefined || endIndex != undefined) {
+      this.#orderedMessageOffset = this.#orderedMessageOffset.slice(startIndex ?? 0, endIndex);
+    }
+  }
+
+  #getSortTime(): bigint {
+    if (
+      this.#orderedMessageOffset != undefined &&
+      this.#orderedMessageOffset.length > 0 &&
+      this.#messageOffsetIndex < this.#orderedMessageOffset.length
+    ) {
+      return this.#orderedMessageOffset[this.#messageOffsetIndex]![0];
+    } else {
```

Comment:
nit: You can drop this _else_ since the above _if_ returns. Consider adding a comment above the _if_ to help clarify a rather complex conditional expression.

Reply:
I don't understand this comment. If message indices haven't been loaded yet then `this.#orderedMessageOffset` is undefined and the `else` branch is taken.

Reply:
You don't have to change it if you prefer this - my comment was a code style nit that because the _if_ condition returns, you can drop the _else_ block and put its return under the _if_. To me that makes the code read more like "here's the normal return path" but if these things are true we might return earlier.

```
    if (
      this.#orderedMessageOffset != undefined &&
      this.#orderedMessageOffset.length > 0 &&
      this.#messageOffsetIndex < this.#orderedMessageOffset.length
    ) {
      return this.#orderedMessageOffset[this.#messageOffsetIndex]![0];
    {

    return this.#reverse ? this.chunkIndex.messageEndTime : this.chunkIndex.messageStartTime;
```